src/shared/pkcs11-util.c

```diff
@@ -698,6 +745,127 @@ int pkcs11_token_decrypt_data(
         return 0;
 }
 
+static inline void pkcs11_token_destroy_session_object(
+                CK_FUNCTION_LIST *m,
+                CK_SESSION_HANDLE session,
+                CK_OBJECT_HANDLE handle) {
+        assert(m);
+
+        CK_RV rv = m->C_DestroyObject(session, handle);
+        if (rv != CKR_OK)
+                log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to destroy object, ignoring.");
+}
+
+#if HAVE_OPENSSL
+int pkcs11_token_ecdh_derive_shared_secret(
+                CK_FUNCTION_LIST *m,
+                CK_SESSION_HANDLE session,
+                CK_OBJECT_HANDLE object,
+                EVP_PKEY *peer_key,
+                uint8_t *ret_shared_secret,
+                size_t *ret_shared_secret_len) {
+
+        assert(m);
+        assert(peer_key);
+        assert(ret_shared_secret_len);
+
+        _cleanup_(freep) uint8_t *buffer = NULL;
+        size_t shared_secret_len = 0;
+        size_t public_data_len = 0;
+        const CK_BBOOL yes = CK_TRUE;
+        const CK_BBOOL no = CK_FALSE;
+        CK_OBJECT_HANDLE shared_secret_handle = CK_INVALID_HANDLE;
+        const CK_OBJECT_CLASS shared_secret_class = CKO_SECRET_KEY;
+        const CK_KEY_TYPE shared_secret_type = CKK_GENERIC_SECRET;
+        CK_RV rv;
+        CK_ECDH1_DERIVE_PARAMS params = {
+                .kdf = CKD_NULL,
+                .pSharedData = NULL,
+                .ulSharedDataLen = 0,
+        };
+        CK_ATTRIBUTE shared_secret_template[] = {
+                {CKA_TOKEN, (void *)&no, sizeof(no)},
+                {CKA_CLASS, (void *)&shared_secret_class, sizeof(shared_secret_class)},
+                {CKA_KEY_TYPE, (void *)&shared_secret_type, sizeof(shared_secret_type)},
+                {CKA_VALUE_LEN, (void *)&shared_secret_len, sizeof(shared_secret_len)},
+                {CKA_SENSITIVE, (void *)&no, sizeof(no)},
+                {CKA_EXTRACTABLE, (void *)&yes, sizeof(yes)},
+                {CKA_ENCRYPT, (void *)&yes, sizeof(yes)},
+                {CKA_DECRYPT, (void *)&yes, sizeof(yes)}
+        };
+        CK_ATTRIBUTE shared_secret_attr = {
+                .type = CKA_VALUE,
+        };
+        CK_MECHANISM mechanism = {
+                .mechanism = CKM_ECDH1_DERIVE,
+                .pParameter = &params,
+                mechanism.ulParameterLen = sizeof(CK_ECDH1_DERIVE_PARAMS)
```

Comment:
this doesnt look right: this is a line with side-effect on the struct itself. weird.

i figure you just want to say `.ulParameterLen` here...